File: src/main/java/org/hazelcast/cache/CachingService.java
package org.hazelcast.cache;

import com.hazelcast.map.IMap;
import org.springframework.data.domain.Sort;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

public class CachingService {

    private final IMap<Long, Person> cache;
    private final PersonRepository repository;

    public CachingService(IMap<Long, Person> cache, PersonRepository repository) {
        this.repository = repository;
        this.cache = cache;
    }

    Mono<Person> findById(Long id) {
        var person = cache.get(id);
        if (person == null) {
            System.out.println("Person with id " + id + " not found in cache");
            var mono = repository.findById(id);
            return mono.doOnSuccess(p -> {
                cache.put(id, p);
                System.out.println("Person with id " + id + " put in cache");
            });
        } else {
            System.out.println("Person with id " + id + " found in cache");
            return Mono.just(person);
        }
    }

    Flux<Person> findAll(Sort sort) {
        return repository
                .findAll(sort)
                .doOnNext(p -> cache.put(p.getId(), p));
    }
}

File: src/main/java/org/hazelcast/cache/CachingService.kt
package org.hazelcast.cache

import com.hazelcast.map.IMap
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.future.await
import org.springframework.data.domain.Sort
import reactor.util.Loggers


class CachingService(private val cache: IMap<Long, Person>, private val repository: PersonRepository) {

    private val logger = Loggers.getLogger(CachingService::class.java)

    suspend fun findById(id: Long) = cache.getAsync(id).await()
        .also {
            if (it == null) logger.info("Person with id $id not found in cache")
            else logger.info("Person with id $id found in cache")
        } ?: repository.findById(id)
            ?.also {
                cache.putAsync(it.id, it)
                logger.info("Person with id $id put in cache")
            }

    suspend fun findAll(sort: Sort) = repository
        .findAll(sort)
        .also { flow ->
            flow.collect {
                cache.putAsync(it.id, it)
            }
        }
}
